File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let ptr = head;
    let target = head;
    let distance = 0;
    while(ptr != null) {
        ptr = ptr.next;
        if(distance == n + 1) {
            target = target.next;
        }
        else {
            distance++;
        }
    }
    if(distance != n + 1) {
        return head.next;
    }
    if(!target.next) return null;
    target.next = target.next.next;
    return head;
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* ptr = head;
        int total = 0;
        while(ptr != nullptr) {
            ptr = ptr -> next;
            total++;
        }
        if(total == 1) {
            return n == 1 ? nullptr : head;
        }
        ptr = head;
        int curr = 1;
        int target = total - n;
        if(target == 0) {
            return head -> next;
        }
        while(curr < target) {
            ptr = ptr->next;
            curr++;
        }
        ptr->next = ptr->next->next;
        return head;
    }
};